src/circular_buffer.jl

```diff
@@ -1,44 +1,50 @@
-
 """
 New items are pushed to the back of the list, overwriting values in a circular fashion.
 """
-type CircularBuffer{T} <: AbstractVector{T}
+type CircularBuffer{T, N} <: AbstractArray{T, N}
     capacity::Int
     first::Int
-    buffer::Vector{T}
+    buffer::Array{T, N}
+    used::Int
+    colons::Vector{Colon}
+
+    function CircularBuffer(capacity::Int, others_dims...)
+        a = Array{T, N}(capacity, others_dims...)
+        colons = Vector{Colon}(N - 1)
+        new(capacity, 1, a, 0, colons)
+    end
+end
 
-    CircularBuffer(capacity::Int) = new(capacity, 1, T[])
+function CircularBuffer(T, capacity::Int, others_dims...)
+    N = length(others_dims) + 1
+    CircularBuffer{T,N}(capacity::Int, others_dims...)
 end
 
 function _buffer_index(cb::CircularBuffer, i::Int)
     n = length(cb)
     if i < 1 || i > n
         throw(BoundsError("CircularBuffer out of range. cb=$cb i=$i"))
     end
-    idx = cb.first + i - 1
-    if idx > n
-        idx - n
-    else
-        idx
-    end
+    idx = mod1(cb.first + i - 1, n)
 end
 
 function Base.getindex(cb::CircularBuffer, i::Int)
-    cb.buffer[_buffer_index(cb, i)]
+    cb.buffer[_buffer_index(cb, i), cb.colons...]
```

Comment:
This seems suboptimal. If you're set on going down the Multidim path, maybe this should be a generated function and you can remove the colons field?